File: src/test/java/com/growingio/kafka/BinarySearchTest.java
package com.growingio.kafka;

import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.junit.Test;
import static org.junit.Assert.assertEquals;

import java.util.*;

/**
 * Created by foolchi on 01/08/15.
 * Unit test
 */
public class BinarySearchTest {

    @Test
    public void testBinarySearch() {
        String topic = "binary_search_test_" + System.currentTimeMillis();

        // create data
        long message_size = 1000;
        Properties props = new Properties();
        props.put("bootstrap.servers", "localhost:9092");
        props.put("acks", "1");
        props.put("key.serializer", "org.apache.kafka.common.serialization.StringSerializer");
        props.put("value.serializer", "org.apache.kafka.common.serialization.StringSerializer");

        KafkaProducer<String, String> producer = new KafkaProducer<String, String>(props);
        Map<String, Long> msgOffsets = new HashMap<String, Long>();

        for (long i = 0; i < message_size; i++) {
            String msg = System.currentTimeMillis() + "";
            producer.send(new ProducerRecord<String, String>(topic, msg));
            msgOffsets.put(msg, i);
            try {
                Thread.sleep(1);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }

        // 验证查找结果
        String[] brokers = new String[]{"localhost"};
        int port = 9092;
        KafkaBinarySearch binarySearch = new KafkaBinarySearch(topic, brokers, port);
        Iterator<Map.Entry<String, Long>> iterator = msgOffsets.entrySet().iterator();
        while (iterator.hasNext()) {
            Map.Entry entry = iterator.next();
            String key = (String) entry.getKey();
            long offset = (Long) entry.getValue();
            assertEquals("Wrong offset: " + key, offset, binarySearch.search(new TimestampComparator(key)));
        }
    }

    @Test
    public void testFuzzyBinarySearch() {
        String topic = "binary_search_test_" + System.currentTimeMillis();
        int disturbance = 10; // 随机扰动范围
        Random r = new Random(System.currentTimeMillis());

        // create data
        long message_size = 100;
        Properties props = new Properties();
        props.put("bootstrap.servers", "localhost:9092");
        props.put("acks", "1");
        props.put("key.serializer", "org.apache.kafka.common.serialization.StringSerializer");
        props.put("value.serializer", "org.apache.kafka.common.serialization.StringSerializer");

        KafkaProducer<String, String> producer = new KafkaProducer<String, String>(props);
        Map<String, Long> msgOffsets = new HashMap<String, Long>();

        for (long i = 0; i < message_size; ) {
            long currentTime = System.currentTimeMillis() + r.nextInt(disturbance);
            String msg = currentTime + "";
            if (!msgOffsets.containsKey(msg)) {
                producer.send(new ProducerRecord<String, String>(topic, msg));
                msgOffsets.put(msg, i);
                i++;
            }
            try {
                Thread.sleep(1);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }

        // 验证查找结果
        String[] brokers = new String[]{"localhost"};
        int port = 9092;

        KafkaBinarySearch binarySearch = new KafkaBinarySearch(topic, brokers, port);
        Iterator<Map.Entry<String, Long>> iterator = msgOffsets.entrySet().iterator();
        while (iterator.hasNext()) {
            Map.Entry entry = iterator.next();
            String key = (String) entry.getKey();
            long offset = (Long) entry.getValue();
//            assertEquals("Wrong offset" + key, offset, binarySearch.fuzzySearch(new TimestampFuzzyComparator(key, disturbance)));
        }
    }

}

File: src/test/scala/com/growingio/kafka/BinarySearchTest.scala
package com.growingio.kafka

import collection.JavaConversions._
import org.apache.kafka.clients.producer.KafkaProducer
import org.apache.kafka.clients.producer.ProducerRecord
import org.junit.Test
import org.junit.Assert.assertEquals

/**
 * Created by foolchi on 04/08/15.
 * Unit test
 */

class BinarySearchTest {

  @Test
  def testBinarySearch() = {
    val topic = "binary_search_test_" + System.currentTimeMillis

    // create data
    val message_size = 1000
    val props = Map(
      "bootstrap.servers" -> "localhost:9092",
      "acks" -> "1",
    "key.serializer" -> "org.apache.kafka.common.serialization.StringSerializer",
      "value.serializer" -> "org.apache.kafka.common.serialization.StringSerializer"
    )
    val producer = new KafkaProducer[String, String](props)
    val msgOffsets = new scala.collection.mutable.HashMap[String, Long]()

    for (i <- 0 until message_size) {
      val msg = System.currentTimeMillis().toString
      producer.send(new ProducerRecord[String, String](topic, msg))
      msgOffsets(msg) = i
      Thread.sleep(1)
    }

    val broker = "localhost"
    val port = 9092
    val binarySearch = new KafkaBinarySearch(topic, broker, port)

    msgOffsets.foreach{
      case (key, value) =>
      assertEquals("Wrong offset: " + key, value, binarySearch.search(new TimestampComparator(key)))
    }
  }

  @Test
  def testFuzzyBinarySearch() = {
    val topic = "binary_search_test_" + System.currentTimeMillis
    val disturbance = 10 // 随机扰动范围
    val r = scala.util.Random

    // create data
    val message_size = 100
    val props = Map(
      "bootstrap.servers" -> "localhost:9092",
      "acks" -> "1",
      "key.serializer" -> "org.apache.kafka.common.serialization.StringSerializer",
      "value.serializer" -> "org.apache.kafka.common.serialization.StringSerializer"
    )
    val producer = new KafkaProducer[String, String](props)
    val msgOffsets = new scala.collection.mutable.HashMap[String, Long]()
    var i = 0

    while (i < message_size) {
      val msg = (System.currentTimeMillis + r.nextInt(disturbance)).toString
      if (!msgOffsets.contains(msg)) {
        producer.send(new ProducerRecord[String, String](topic, msg))
        msgOffsets(msg) = i
        i += 1
      }
      Thread.sleep(1)
    }

    val broker = "localhost"
    val port = 9092
    val binarySearch = new KafkaBinarySearch(topic, broker, port)

    msgOffsets.foreach{
      case (key, value) =>
        assertEquals("Wrong offset: " + key, value, binarySearch.fuzzySearch(new TimestampFuzzyComparator(key, disturbance)))
    }
  }

}
